File: instacapture/src/main/java/com/tarek360/instacapture/ActivityReferenceManager.java
package com.tarek360.instacapture;

import android.app.Activity;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;

/**
 * Created by tarek on 5/17/16.
 */
public final class ActivityReferenceManager {

    @Nullable
    private WeakReference<Activity> mActivity;

    public void setActivity(@NonNull final Activity activity) {
        this.mActivity = new WeakReference<>(activity);
    }

    @Nullable
    public Activity getValidatedActivity() {
        if (mActivity == null) {
            return null;
        }

        final Activity activity = mActivity.get();
        if (!isActivityValid(activity)) {
            return null;
        }

        return activity;
    }

    private boolean isActivityValid(@Nullable final Activity activity) {
        if (activity == null) {
            return false;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return !activity.isFinishing() && !activity.isDestroyed();
        } else {
            return !activity.isFinishing();
        }
    }
}

File: instacapture/src/main/java/com/tarek360/instacapture/ActivityReferenceManager.kt
package com.tarek360.instacapture

import android.app.Activity
import android.os.Build

import java.lang.ref.WeakReference

/**
 * Created by tarek on 5/17/16.
 */
class ActivityReferenceManager {

    private var activity: WeakReference<Activity>? = null

    fun setActivity(activity: Activity) {
        this.activity = WeakReference(activity)
    }

    val validatedActivity: Activity?
        get() {
            if (activity == null) {
                return null
            }

            val activity = activity!!.get()
            if (!isActivityValid(activity)) {
                return null
            }

            return activity
        }

    private fun isActivityValid(activity: Activity?): Boolean {
        if (activity == null) {
            return false
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return !activity.isFinishing && !activity.isDestroyed
        } else {
            return !activity.isFinishing
        }
    }
}
